feat(catalog-client): Support generating a front end client based on an OpenAPI spec

sennyeya opened the pull request:
## Hey, I just made a Pull Request!

## Summary
Another step for #2566. There's a lot of code with this PR, the basic idea is to support using `openapi-generator` with some overrides to create a "closer to the spec" API client.

## What's in this PR?
The primary work is in the `packages/repo-tools/templates` which is a set of overrides to the Mustache templates that `openapi-generator`'s `typescript` generator uses under the hood. There are some unfortunate requirements there, namely the `.model.ts` and `.client.ts`. My goal here was to reuse the upstream template as much as possible.

I also adjusted the current CatalogClient to use the generated code under the hood instead of creating URLs and requests. The `packages/catalog-client/src/generated` folder is from the custom generator.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

sennyeya commented:
@freben @taras I was able to get a custom generator set up from `openapi-generator` that resembles most of the CatalogClient today. It's in Java and a lot of the code is based off of [`TypescriptClientCodegen.java`](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptClientCodegen.java). I can take a look at `openapi-generator-plus` but `openapi-generator` is definitely more standard and has support for more cases like server stubs and clients down the line. 

That being said, I have 2 asks,
1. Can you take a look at the output types (in `catalog-client/src/generated/models/*`) and the output API (in `catalog-client/src/generated/api/DefaultApi.ts`) and see if they're helpful? 
2. I pulled in the existing serialization code from their Typescript generator, which makes a lot of the methods _much_ more verbose, does this make sense to add? It feels like it adds a fair amount of safety to our API calls that we didn't have before.

taras commented:
@sennyeya, awesome work on this. Thank you for pushing it forward.

> Can you take a look at the output types (in catalog-client/src/generated/models/*) and the output API (in catalog-client/src/generated/api/DefaultApi.ts) and see if they're helpful?

`generated/api/DefaultAPI.ts` has `Promise<Response>` as return for all methods. This was a problem that I saw previously with `openapi-generator`. We need the return to be typed, but it's not (at least from what I can see).

`catalog-client/src/generated/models/*` is very java-ish. It generates a serializer class for something that could be a type. It doesn't use that class to type the response.

> I pulled in the existing serialization code from their Typescript generator, which makes a lot of the methods much more verbose, does this make sense to add?

I'm not sure, maybe @freben has a better idea.

> I can take a look at `openapi-generator-plus` but `openapi-generator` is definitely more standard and has support for more cases like server stubs and clients down the line.

Just a heads up, I did look at them previously. One of the challenge with those libraries is that their maintainers don't have much time.  I opened a few issues a year ago https://github.com/karlvr/openapi-generator-plus-generators/issues/32 and https://github.com/karlvr/openapi-generator-plus-generators/issues/33. We could add eslint ignore, but just a heads up.

sennyeya commented:
@taras  Thanks!

> generated/api/DefaultAPI.ts has Promise<Response> as return for all methods. This was a problem that I saw previously with openapi-generator. We need the return to be typed, but it's not (at least from what I can see).

This was something that I was trying to hit an in between state with, the CatalogClient currently handles different endpoints separately (see the `requestRequired`, `requestOptional` and `requestIgnored` internal methods). We know the return types in the template but have to decide how to parse them and if we want to handle errors (ideally the errors are pushed to the caller and they handle them).

> catalog-client/src/generated/models/* is very java-ish. It generates a serializer class for something that could be a type. It doesn't use that class to type the response.

Similar to above, I kept the serializer class from their implementation as an easier bridge from their typescript client, we can definitely remove (see `template-openapi-client-plugin/generator/src/main/resources/typescript-backstage/model.mustache`, but figured it would be useful for correctly serializing FormData/bodies.

> Just a heads up, I did look at them previously. One of the challenge with those libraries is that their maintainers don't have much time. I opened a few issues a year ago https://github.com/karlvr/openapi-generator-plus-generators/issues/32 and https://github.com/karlvr/openapi-generator-plus-generators/issues/33. We could add eslint ignore, but just a heads up.

Yeah, that was one of my concerns, that library seems to have just 1 or 2 maintainers vs `openapi-generator` which has _quite_ a few and a wider range of templates. My main issue with it is we have to write Java to customize the template if we want to maintain a custom generator (which I think we should). Not sure where this project sits on adding Java code, which is probably a question for @freben.

sennyeya commented:
Did a little more investigation on the generated types and support, the `openapi-generator` typescript generator doesn't have support for `oneOf` or `allOf`, or `additionalProperties: true`. That's causing an issue from the `lint` step where the `Entity` from `@backstage/catalog-models` and the autogenerated `Entity` from the spec don't match. (There's work in the future to combine the two of those). Basically, because `openapi-generator` doesn't create a `[key:string]: any;` in addition to known properties on the interface, Typscript can't merge the types and we fail validation.

I looked into `autorest` to see if it handles those cases better and it handles them as expected. I'm not at all versed in `ts-morph`, but will start taking a look this week as `openapi-generator` _probably_ isn't the direction anymore.

taras commented:
Thank you for continuing to dig through this. Excuted to see what autorest implementation looks like.

sennyeya commented:
@taras It looks like `autorest` is super tied to their `@azure` HTTP library and auth implementations, were you able to create a custom plugin? If so, do you have a library/implementation you could share. I have an open question to them, but poking around their implementation, it's looking like we'd have to create/fork their implementation of all layers (TypeSpec->RLCModel and RLCModel->code) instead of just the top level.

taras commented:
@sennyeya, I didn't get very far. I looked at how much work would involve making a new template, realized that I probably wouldn't have enough time to see it through, and moved on. At the time, I thought we'd need to create a reference client and work backward to create a template. I'm not sure about the plugins, but your assessment sounds right.

sennyeya commented:
This PR is still a WIP, `autorest` didn't work out, and so this will require upstream work in `openapi-generator` to support `additionalProperties`. Timeline for that unknown.

sennyeya commented:
@taras @Rugvip @tudi2d I finally got some time to take a look at this one again. Looks like there has been some great work on the openapi-generator side that unblocks _most_ of our use case 😄 ! `additionalProperties: true` still isn't supported, buuut we can do `additionalProperties: type: string` or whatnot and get around it, which makes the types line up between `catalog-model` and the generated types! 🎉 

Couple of other things that I added as part of the recent changes,
1. Use of URI templating for creating the query parameters. This makes it substantially less painful to create the query parameters and their encoding/whatnot. The out of the box version didn't do this and it might be a cool upstream change to make in time.
2. Using `/catalog` instead of just `/` for the server to give the client a pluginId to work with in the discoveryApi. I assume this will be controversial and open to other ideas on how to retrieve it from the client/template.
3. I moved the templates for the new client plugins directly into `repo-tools`, following a similar idea as the create plugin process in `cli`. It might make sense to make this a part of `cli`, but I've been mostly in `repo-tools` so started there.
4. There are still a few install weirdnesses I have to document fully, the main one being that the output of `schema openapi generate-client` doesn't fully create a new plugin and instead adds files to a `src/generated` folder with the assumption that you will write the wrappers yourself.

freben commented:
seems there's a couple of conflicts.

sennyeya commented:
@freben A good problem to have, _too_ many new OpenAPI commands. Fixed!

freben commented:
Sorry I haven't reviewed yet - but the changesets are missing!

sennyeya commented:
@freben Added!

Rugvip commented on packages/catalog-client/src/generated/apis/DefaultApi.client.ts:
```diff
@@ -0,0 +1,828 @@
+      method: 'DELETE',
+      body,
+    });
+    if (!response.ok) {
+      throw new ResponseError(response);
+    }
+    if (isCodeInRange('204', response.status)) {
+      return (await response.text()) as any;
+    }
+    if (isCodeInRange('400', response.status)) {
+      return (await response.text()) as any;
+    }
+    if (isCodeInRange('0', response.status)) {
+      return (await response.text()) as any;
+    }
+    return await response.json();
```
Just a note that I think we could think a about exactly what to return here. I'm feeling that a `ResponseError.fromResponse(res)` *should* make sense here but maybe it doesn't. Also feeling we likely want to return the response object as is rather than the body. There may be situations we want to access headers and whatnot.

sennyeya replied:
I like the idea of switching to `ResponseError.fromResponse(res)`. I remembered the issue I was having though, basically, `validate-entity` has a weird response shape of `{errors : Error[]}` that doesn't match the generic error response. Open to suggestions here, my thinking is either to add a `rawBody` property to `ResponseError` (probably better) and then parse from there or adjust the `validate-entity` return value (less good)

Rugvip commented on packages/repo-tools/openapitools.json:
Another one here where I wonder if we can skip it

Rugvip commented on packages/repo-tools/src/commands/index.ts:
```diff
@@ -60,6 +64,16 @@ function registerSchemaCommand(program: Command) {
     .action(lazy(() => import('./openapi/test/init').then(m => m.default)));
+
+  openApiCommand
+    .command('generate-client')
```
Yep think this makes sense. We could potentially consider having the other one be `generate-server` and ofc if possible it might be nice to consolidate into a single `generate` command

Rugvip commented on packages/repo-tools/src/commands/openapi/client/generate.ts:
```diff
@@ -0,0 +1,73 @@
+  );
+
+  await exec(
+    `cd ${cliPaths.ownDir} && yarn openapi-generator-cli generate -i "${resolvedOpenapiPath}" -o "${resolvedOutputDirectory}" -g typescript -c "templates/typescript-backstage.yaml"`,
```
Use the `cwd` option rather than `cd`. I'd say it's also nicer to look up the `openapi-generator-cli` bin path rather than calling it via `yarn`. Prefer using `execFile` or similar as well, one where you can pass an array of parameters rather than a full command string. Helps avoid escaping issues.

Rugvip reviewed (changes requested):
Nice 😎

sennyeya commented on packages/catalog-client/src/generated/.gitignore:
As this gets added to the package level `dist` folder and not this folder, will just remove this file. The package-level `dist` is already ignored.

sennyeya commented on packages/catalog-client/src/generated/.openapi-generator/FILES:
This and the next file should also be removed.

sennyeya commented on packages/catalog-client/src/generated/models/ErrorError.model.ts:
Naming is weird here

sennyeya replied:
`Error` clashes with the TS name, not an easy fix to make this a better name.

sennyeya commented on packages/repo-tools/src/commands/openapi/client/generate.ts:
```diff
@@ -0,0 +1,89 @@
+    OPENAPI_IGNORE_FILES.join('\n'),
+  );
+
+  let opticLocation = '';
```
Update variable name here.

sennyeya commented on packages/repo-tools/src/commands/openapi/client/generate.ts:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  await exec(
+    `yarn openapi-generator-cli generate -i "${resolvedOpenapiPath}" -o "${resolvedOutputDirectory}" -g typescript -c "templates/typescript-backstage.yaml"`,
```
rework to use above location var. execFile wasn't working as expected here and will need more tweaking.

sennyeya commented on plugins/catalog-backend/src/schema/openapi.generated.ts:
```diff
@@ -248,12 +252,13 @@ export const spec = {
         properties: {},
         description: 'A type representing all allowed JSON object values.',
-        additionalProperties: true,
+        additionalProperties: {},
```
These are functionally identical, a weird byproduct of using JSON Schema :)

sennyeya commented on plugins/catalog-backend/src/schema/openapi.generated.ts:
```diff
@@ -403,7 +400,6 @@ export const spec = {
         required: ['metadata', 'kind', 'apiVersion'],
         description:
           "The parts of the format that's common to all versions/kinds of entity.",
-        additionalProperties: true,
```
@eide Thoughts on the above? https://github.com/backstage/backstage/blob/b7540d3fd702a653eaf43d858b16d39f47ff348b/packages/catalog-model/src/schema/Entity.schema.json#L28C24-L28C24

sennyeya commented on plugins/catalog-backend/src/schema/openapi.generated.ts:
```diff
@@ -1257,8 +1245,8 @@ export const spec = {
-          '200': {
-            description: 'Ok',
+          '201': {
+            description: '201 response',
```
```suggestion
            description: 'Created',
```

sennyeya commented on packages/catalog-client/src/CatalogClient.ts:
```diff
@@ -434,101 +413,38 @@ export class CatalogClient implements CatalogApi {
-    if (!response.ok) {
-      if (response.status === 404) {
-        return undefined;
+    } catch (e) {
```
Makes sense, I'll revert this. Unfortunate though... I wonder if there's a path to move folks towards the `@backstage/error` package more forcefully. There were a few other usages of custom 4xx errors I saw that would be nice to standardize, [here](https://github.com/backstage/backstage/blob/695984e90b9b75426ffe985abf7bdae0fa9cc1f0/plugins/entity-feedback-backend/src/service/router.ts#L140) and [here](https://github.com/backstage/backstage/blob/695984e90b9b75426ffe985abf7bdae0fa9cc1f0/plugins/tech-insights-backend/src/service/router.ts#L177) among others.

sennyeya replied:
Updated!

Rugvip commented on packages/catalog-client/src/CatalogClient.ts:
```diff
@@ -468,51 +416,21 @@ export class CatalogClient implements CatalogApi {
-      : {};
-    const response = await this.fetchApi.fetch(url, { method, headers });
-
+  private async requestIgnored(response: Response): Promise<void> {
```
I think we could tweak the naming of these now that they're not executing the request and just handling the response. Feeling we can leave that for a followup though

Rugvip commented on packages/cli/config/eslint-factory.js:
```diff
@@ -165,6 +165,23 @@ function createConfig(dir, extraConfig = {}) {
+        rules: {
+          ...tsRules,
+          'no-unused-vars': 'off',
+          'unused-imports/no-unused-imports': 'error',
+          'unused-imports/no-unused-vars': [
```
Hold up, why are these added here? `@typescript-eslint` is what we use for unused vars. In general I feel there's really no point in enabling any additional rules for generated code?

Rugvip replied:
Ah so it trims away things to make `tsc` work? Makes sense then, let's :shipit:

Rugvip reviewed (approved):
NIce! 😁 

I think we're good to ship this and keep iterating. In particular figuring out the error responses, I agree defaulting to `@backstage/errors` would be nice, as well as cleaning up the lint rules and avoid the extra dep.

taras commented:
Amazing work @sennyeya thank you 🙏
